Fix: Cached AccessToken missing realm property in OIDC scenario

bbush915 opened the pull request:
In the case an application is using a separate OIDC-compliant authority and the tid claim is empty, the cached access token does not contain a realm property. One side-effect is that the cached access token will fail the isAccessTokenEntity check, preventing it from being re-used.

bbush915 committed:
fix: Ensure access token hasOwnProperty("realm") when using a separate OIDC-compliant authority.

In the case an application is using a separate OIDC-compliant authority and the tid claim is empty, the cached access token does not contain a realm property. One side-effect is that the cached access token will fail the isAccessTokenEntity check, preventing it from being re-used.

tnorling commented on lib/msal-common/src/response/ResponseHandler.ts:
```diff
@@ -468,7 +468,9 @@ export class ResponseHandler {
                 serverTokenResponse.access_token,
                 this.clientId,
-                idTokenClaims?.tid || authority.tenant,
+                idTokenClaims?.tid ||
+                    authority.tenant ||
+                    Constants.EMPTY_STRING,
```
```suggestion
                idTokenClaims?.tid ||
                    authority.tenant ||
                    "",
```

bbush915 replied:
Updated!

tnorling reviewed (changes requested):
Thanks for the PR! Can you please add a test & generate changefiles by running `npm run beachball:change` from the root of the repo.

bbush915 committed:
feat: Update constant usage and add test for OIDC scenario.

bbush915 commented:
@microsoft-github-policy-service agree [company="Dialexa, an IBM Company"]

bbush915 commented:
@microsoft-github-policy-service agree company="Dialexa, an IBM Company"

bbush915 commented:
@tnorling addressed comment and added a test. The beachball command reported no change files are needed.

tnorling commented:
> @tnorling addressed comment and added a test. The beachball command reported no change files are needed.

Ah I suspect this is because your feature branch is also called dev. Can you try another branch?

bbush915 committed:
Change files

bbush915 commented:
> > @tnorling addressed comment and added a test. The beachball command reported no change files are needed.
> 
> Ah I suspect this is because your feature branch is also called dev. Can you try another branch?

Good call! Sorry for the messy commits, but I got the change files in place!

bbush915 committed:
fix: Use valid client_info.

bbush915 commented:
@tnorling Hello! Is there anything I can do to help this land?

sameerag commented:
@bbush915 I can help land this. Looks like some tests are failing.

bbush915 commented:
@sameerag I validated that tests are all passing locally. Is there anything I can do to troubleshoot the failing E2E tests? It looks like they are all reporting: `Git fetch failed with exit code: 128`

bbush915 committed:
fix: Use claimsTenantId

bbush915 commented:
@tnorling @sameerag following up on the above, I'm not sure why the E2E tests are failing, but can do what is needed to assist. Hoping to get this to land soon, since it is blocking some upgrades.

BboyAkers commented:
Hey cool and awesome former colleagues 🙂! Can this get approved and merged? It's blocking a couple companies and their clients including one I'm helping with 😅🙏🏾. So sorry, for the push and hope I don't come off as rude!

tnorling committed:
Merge branch 'dev' into dev

sameerag committed:
Merge branch 'dev' into dev

sameerag commented:
> Hey cool and awesome former colleagues 🙂! Can this get approved and merged? It's blocking a couple companies and their clients including one I'm helping with 😅🙏🏾. So sorry, for the push and hope I don't come off as rude!

Approved and updated the base branch. Hopefully tests pass now.

BboyAkers commented:
> > Hey cool and awesome former colleagues 🙂! Can this get approved and merged? It's blocking a couple companies and their clients including one I'm helping with 😅🙏🏾. So sorry, for the push and hope I don't come off as rude!
> 
> Approved and updated the base branch. Hopefully tests pass now.

 😭 🙏🏾 ❤️
[![image](https://github.com/AzureAD/microsoft-authentication-library-for-js/assets/11778717/158cf409-ccca-4669-ae9a-d5aa5d4565fb)](https://media.tenor.co/images/3ed84016385a06560957920423fea2f8/raw)
